File: tests/src/main/java/com/github/klee0kai/test/tech/phone/base/ATech.java
package com.github.klee0kai.test.tech.phone.base;

import java.util.LinkedList;

public abstract class ATech {

    private LinkedList<ATechLifecycle> listeners = new LinkedList<>();

    public void subscribe(ATechLifecycle listener) {
        this.listeners.add(listener);
    }

    public void onBuy() {
        for (ATechLifecycle lis : listeners)
            lis.onBuy();
    }

    public void onBroken() {
        for (ATechLifecycle lis : listeners)
            lis.onBroken();
    }

    public void onDrown() {
        for (ATechLifecycle lis : listeners)
            lis.onDrown();
    }


}

File: tests/src/main/java/com/github/klee0kai/test/tech/phone/base/ATech.kt
package com.github.klee0kai.test.tech.phone.base

import java.util.*

abstract class ATech {
    private val listeners = LinkedList<ATechLifecycle>()

    fun subscribe(listener: ATechLifecycle?) {
        this.listeners.add(listener!!)
    }

    fun onBuy() {
        for (lis in listeners) lis.onBuy()
    }

    fun onBroken() {
        for (lis in listeners) lis.onBroken()
    }

    fun onDrown() {
        for (lis in listeners) lis.onDrown()
    }
}
